File: ios/Sources/DatePickerPlugin/DateParser.swift
import Foundation

enum DateParser {
    private static let fallbackFormats = [
        "yyyy-MM-dd'T'HH:mm:ss.SSSXXXXX",
        "yyyy-MM-dd'T'HH:mm:ssXXXXX",
        "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'",
        "yyyy-MM-dd'T'HH:mm:ss'Z'",
        "yyyy-MM-dd'T'HH:mm:ss.SSS",
        "yyyy-MM-dd'T'HH:mm:ss",
        "yyyy-MM-dd'T'HH:mm",
        "yyyy-MM-dd",
        "HH:mm:ss",
        "HH:mm"
    ]

    static func parse(_ value: String?, options: DatePickerOptions) throws -> Date? {
        guard let value, !value.isEmpty else {
            return nil
        }

        if let dateOnly = parseDateOnly(value, options: options) {
            return dateOnly
        }

        if let isoDate = parseISO(value) {
            return isoDate
        }

        if let formatted = parse(value, format: normalizeFormat(options.format), options: options) {
            return formatted
        }

        for format in fallbackFormats {
            if let formatted = parse(value, format: format, options: options) {
                return formatted
            }
        }

        throw DatePickerError.parse("Unable to parse date: \(value)")
    }

    static func format(_ date: Date, options: DatePickerOptions) -> String {
        let formatter = DateFormatter()
        formatter.dateFormat = normalizeFormat(options.format)
        formatter.locale = options.localeValue
        formatter.timeZone = options.timeZoneValue
        return formatter.string(from: date)
    }

    static func normalizeFormat(_ format: String?) -> String {
        guard let format, !format.isEmpty else {
            return DatePickerOptions.defaultFormat
        }

        return format
            .replacingOccurrences(of: "YYYY", with: "yyyy")
            .replacingOccurrences(of: "DD", with: "dd")
            .replacingOccurrences(of: "sss", with: "SSS")
    }

    private static func parse(_ value: String, format: String, options: DatePickerOptions) -> Date? {
        let formatter = DateFormatter()
        formatter.dateFormat = format
        formatter.locale = options.localeValue
        formatter.timeZone = options.timeZoneValue
        formatter.isLenient = false
        return formatter.date(from: value)
    }

    private static func parseISO(_ value: String) -> Date? {
        let fractional = ISO8601DateFormatter()
        fractional.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        if let date = fractional.date(from: value) {
            return date
        }

        let standard = ISO8601DateFormatter()
        standard.formatOptions = [.withInternetDateTime]
        return standard.date(from: value)
    }

    private static func parseDateOnly(_ value: String, options: DatePickerOptions) -> Date? {
        let parts = value.split(separator: "-")
        guard parts.count == 3,
              let year = Int(parts[0]),
              let month = Int(parts[1]),
              let day = Int(parts[2]) else {
            return nil
        }

        var calendar = Calendar(identifier: .gregorian)
        calendar.locale = options.localeValue
        calendar.timeZone = options.timeZoneValue

        var components = DateComponents()
        components.calendar = calendar
        components.timeZone = options.timeZoneValue
        components.year = year
        components.month = month
        components.day = day
        components.hour = 12
        return calendar.date(from: components)
    }
}

enum DatePickerError: Error {
    case parse(String)
}

File: android/src/main/java/app/capgo/datepicker/DateParser.java
package app.capgo.datepicker;

import java.text.ParseException;
import java.text.ParsePosition;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.Locale;
import java.util.TimeZone;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

final class DateParser {

    private static final Pattern DATE_ONLY = Pattern.compile("^(\\d{4})-(\\d{2})-(\\d{2})$");
    private static final String[] FALLBACK_FORMATS = {
        "yyyy-MM-dd'T'HH:mm:ss.SSSXXX",
        "yyyy-MM-dd'T'HH:mm:ss.SSSX",
        "yyyy-MM-dd'T'HH:mm:ssXXX",
        "yyyy-MM-dd'T'HH:mm:ssX",
        "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'",
        "yyyy-MM-dd'T'HH:mm:ss'Z'",
        "yyyy-MM-dd'T'HH:mm:ss.SSS",
        "yyyy-MM-dd'T'HH:mm:ss",
        "yyyy-MM-dd'T'HH:mm",
        "yyyy-MM-dd",
        "HH:mm:ss",
        "HH:mm"
    };

    private DateParser() {}

    static Date parse(String value, DatePickerOptions options) throws ParseException {
        if (value == null || value.isEmpty()) {
            return null;
        }

        Matcher dateOnly = DATE_ONLY.matcher(value);
        if (dateOnly.matches()) {
            return dateOnly(
                Integer.parseInt(dateOnly.group(1)),
                Integer.parseInt(dateOnly.group(2)),
                Integer.parseInt(dateOnly.group(3)),
                options
            );
        }

        String format = normalizeFormat(options.format);
        Date formatted = parseWithFormat(value, format, options);
        if (formatted != null) {
            return formatted;
        }

        for (String fallback : FALLBACK_FORMATS) {
            Date parsed = parseWithFormat(value, fallback, options);
            if (parsed != null) {
                return parsed;
            }
        }

        throw new ParseException("Unable to parse date: " + value, 0);
    }

    static String format(Date date, DatePickerOptions options) {
        if (date == null) {
            return null;
        }

        SimpleDateFormat dateFormat = new SimpleDateFormat(normalizeFormat(options.format), options.localeValue());
        dateFormat.setTimeZone(options.timeZoneValue());
        return dateFormat.format(date);
    }

    static String normalizeFormat(String format) {
        if (format == null || format.isEmpty()) {
            return DatePickerOptions.DEFAULT_FORMAT;
        }

        return format.replace("YYYY", "yyyy").replace("DD", "dd").replace("sss", "SSS");
    }

    private static Date parseWithFormat(String value, String format, DatePickerOptions options) {
        SimpleDateFormat dateFormat = new SimpleDateFormat(format, options.localeValue());
        dateFormat.setLenient(false);
        dateFormat.setTimeZone(options.timeZoneValue());

        ParsePosition position = new ParsePosition(0);
        Date parsed = dateFormat.parse(value, position);
        if (parsed == null || position.getIndex() != value.length()) {
            return null;
        }

        return parsed;
    }

    private static Date dateOnly(int year, int month, int day, DatePickerOptions options) {
        Calendar calendar = Calendar.getInstance(options.timeZoneValue(), options.localeValue());
        calendar.clear();
        calendar.set(Calendar.YEAR, year);
        calendar.set(Calendar.MONTH, month - 1);
        calendar.set(Calendar.DAY_OF_MONTH, day);
        calendar.set(Calendar.HOUR_OF_DAY, 12);
        return calendar.getTime();
    }
}
